datacube/_celery_runner.py

```diff
@@ -0,0 +1,259 @@
+    if password == '':
+        password = get_redis_password(generate_if_missing=False)
+
+    set_address(host, port, password=password)
+
+    argv = ['worker', '-A', 'datacube._celery_runner']
+    if nprocs is not None:
+        argv.extend(['-c', str(nprocs)])
+
+    app.worker_main(argv)
+
+
+def get_redis_password(generate_if_missing=False):
+    from .utils import write_user_secret_file, slurp, gen_password
+
+    REDIS_PASSWORD_FILE = '.datacube-redis'
+
+    password = slurp(REDIS_PASSWORD_FILE, in_home_dir=True)
+    if password is not None:
+        return password
+
+    if generate_if_missing:
+        password = gen_password(12)
+        write_user_secret_file(password, REDIS_PASSWORD_FILE, in_home_dir=True)
+
+    return password
+
+
+class CeleryExecutor(object):
+    def __init__(self, host=None, port=None, password=None):
+        # print('Celery: {}:{}'.format(host, port))
+        self._shutdown = None
+
+        if port or host or password:
+            if password == '':
+                password = get_redis_password(generate_if_missing=True)
+
+            set_address(host if host else 'localhost',
+                        port if port else 6379,
+                        password=password)
+
+        host = host if host else 'localhost'
+        port = port if port else 6379
+
+        if not check_redis(host, port, password):
+            if host in ['localhost', '127.0.0.1']:
+                self._shutdown = launch_redis(port if port else 6379, password=password)
+            else:
+                raise IOError("Can't connect to redis server @ {}:{}".format(host, port))
+
+    def __del__(self):
+        if self._shutdown:
+            app.control.shutdown()
+            sleep(1)
+            self._shutdown()
+
+    def __repr__(self):
+        return 'CeleryRunner'
+
+    def submit(self, func, *args, **kwargs):
+        return submit_cloud_pickled_function(func, *args, **kwargs)
+
+    def map(self, func, iterable):
+        return [self.submit(func, data) for data in iterable]
+
+    @staticmethod
+    def get_ready(futures):
+        completed = []
+        failed = []
+        pending = []
+        for f in futures:
+            if f.ready():
+                if f.failed():
+                    failed.append(f)
+                else:
+                    completed.append(f)
+            else:
+                pending.append(f)
+        return completed, failed, pending
+
+    @staticmethod
+    def as_completed(futures):
+        while len(futures) > 0:
+            pending = []
+
+            for promise in futures:
+                if promise.ready():
+                    yield promise
+                else:
+                    pending.append(promise)
+
+            if len(pending) == len(futures):
+                # If no change detected sleep for a bit
+                # TODO: this is sub-optimal, not sure what other options are
+                #       though?
+                sleep(0.1)
```

Comment:
It is a shame that `celery.ResultSet` doesn't have a `collect()` function like `AsyncResult`  http://docs.celeryproject.org/en/latest/reference/celery.result.html#celery.result.AsyncResult.collect